vpc-gateway-endpoints.tf

```diff
@@ -0,0 +1,86 @@
+# ---------------------------------------------------------------------------------------------------------------------
+# VPC GATEWAY ENDPOINTS
+#
+# A VPC endpoint enables you to privately connect your VPC to supported AWS services and VPC endpoint services powered
+# by AWS PrivateLink without requiring an internet gateway, NAT device, VPN connection,
+# or AWS Direct Connect connection. Instances in your VPC do not require public IP addresses to communicate with
+# resources in the service. Traffic between your VPC and the other service does not leave the Amazon network.
+# ---------------------------------------------------------------------------------------------------------------------
+
+# ---------------------------------------------------------------------------------------------------------------------
+# VPC Gateway Endpoint for S3
+# ---------------------------------------------------------------------------------------------------------------------
+
+data "aws_vpc_endpoint_service" "s3" {
+  count = var.create && var.enable_s3_endpoint ? 1 : 0
+
+  service = "s3"
+}
+
+resource "aws_vpc_endpoint" "s3" {
+  count = var.create && var.enable_s3_endpoint ? 1 : 0
+
+  vpc_id       = aws_vpc.vpc[0].id
+  service_name = data.aws_vpc_endpoint_service.s3[0].service_name
+  tags         = var.tags
+}
+
+resource "aws_vpc_endpoint_route_table_association" "s3_public" {
+  count = var.create && var.enable_s3_endpoint && length(aws_subnet.public) > 0 ? 1 : 0
+
+  vpc_endpoint_id = aws_vpc_endpoint.s3[0].id
+  route_table_id  = aws_route_table.public[0].id
+}
+
+resource "aws_vpc_endpoint_route_table_association" "s3_private" {
+  for_each = var.create && var.enable_s3_endpoint ? aws_subnet.private : {}
+
+  vpc_endpoint_id = aws_vpc_endpoint.s3[0].id
+  route_table_id  = aws_route_table.private[each.key].id
+}
+
+resource "aws_vpc_endpoint_route_table_association" "s3_intra" {
+  for_each = var.create && var.enable_s3_endpoint ? aws_subnet.intra : {}
+
+  vpc_endpoint_id = aws_vpc_endpoint.s3[0].id
+  route_table_id  = aws_route_table.intra[each.key].id
+}
```

Comment:
TL;DR: i want multiple single-purpose VPC endpoints granting least needed privileges for every service.

my 2cents here and the issues i see:
imagine this (made up) real (yes, made it up, but could be real) world example use case:

General: As a developer, I want to use VPC Endpoints to secure communication between my App running in a VPC and AWS services. 
More detailed: As a developer, I have an application where the frontend (running in the private subnet) fetches data from S3, renders it and returns it to the client. The data in S3 is created by a cron (running in intra subnet) and pushed to S3 to be available for the frontend.

i want both, the frontend and the cron to access S3 via VPC endpoints. the frontend should only read the bucket and the cron should only write.

- I want to create an endpoint (A) that grants the frontend read-only.
- I want to create an endpoint (B) that grants the cron write-only.
- I want a bucket policy (in cross-account scenario) that ensures that those VPC endpoints can only do what they are supposed to to..

so what setup  solves this:
- one s3 endpoint routed from intra subnet granting write-only access
- one s3 endpoint routed from private subnet granting read-only
- some policies making sure the access is limited to those least needed privileges

Reply:
Okay, my 2 cents:

Since we'd like to keep our modules as flexible as possible. We could do the following:
Gateway Endpoints should always be enabled for  private and intra subnets ( how about public ? ) but can we disabled if needed ?

In addition to that, I would make the Interface Endpoints fully configurable. Would it be enough to do that on a subnet type level? How about something like that:

```hcl
  public_subnets = [
    {
      availability_zone = "us-east-1a",
      cidr_block        = "10.0.64.0/21",
    },
    {
      availability_zone = "us-east-1a",
      cidr_block        = "10.0.80.0/21",
    }....
  ]

public_subnet_interface_endpoints = [
{
  service = "sagemaker.runtime" # a pre-defined map of services to choose from would be a good idea here
  security_group_ids   = []
  private_dns_enabled = true
  policy = "somepolicyarn"
custom_tags = {foo = "faa"}
}
]

```

Also, It might be a reasonable approach to split this PR into two ( one per endpoint type ).

Looking forward to getting your feedback on this.